design/pool-configuration.md

```diff
@@ -0,0 +1,533 @@
+
+```go
+type bgpAdvertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+  AggregationLength *int
+  LocalPref         *uint32
+  Communities       []string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
+metadata:
+  name: bgpadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+    peers:
+      - peer1
+      - peer2
+    - communities:
+        - 65535:65282
+    aggregationLength: 32
+    localPref: 100
+```
+
+Open question: I would like to preserve not having to specify all the ips / peers in order to match the current "ease of configuration" of metallb. One possible way is to assume that all the pools are advertised to all the peers if no selection is specified, the other one is to provide an example using the label selector.
+
+#### L2 Advertisement
+
+The `L2Advertisement` entity represents the intention to advertise a given IP via L2. Currently, there are no knobs related to L2 advertisement.
+
+```go
+type l2Advertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
+}
+```
+
+Example CRD:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: lsadvertisement
+  namespace: metallb-system
+spec:
+    pools:
+      - pool1
+      - pool2
+```
+
+### BGP Peer
+
+The `BgpPeer` entity represents an external BGP peer we want to establish a session with. It contains the properties of the sessions, and it's a 1:1 mapping of the current bgp peer structure without the node selector. We also add the `bfdProfile` structure here, as it is part of the session definition.
+
+```go
+type bgpPeer struct {
+  Name              string
+  MyASN         uint32
+  ASN           uint32
+  Addr          string
+  SrcAddr       string
+  Port          uint16
+  HoldTime      string
+  KeepaliveTime string
+  RouterID      string
+  Password      string
+  BFDProfile    string
+  EBGPMultiHop  bool
+}
+
+type bfdProfile struct {
+  Name             string  
+  ReceiveInterval  *uint32 
+  TransmitInterval *uint32 
+  DetectMultiplier *uint32 
+  EchoInterval     *uint32 
+  EchoMode         bool    
+  PassiveMode      bool    
+  MinimumTTL       *uint32 
+}
+```
+
+Example CRDs:
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: BGPPeer
+metadata:
+  name: peer-bfd
+  namespace: metallb-system
+spec:
+  peerAddress: 10.0.0.2
+  peerASN: 64501
+  myASN: 64500
+  routerID: 10.10.10.10
+  peerPort: 1
+  holdTime: "10s"
+  sourceAddress: 1.1.1.1
+  bfdProfile: bfdprofilefull
+
+apiVersion: metallb.io/v1beta1
+kind: BFDProfile
+metadata:
+  name: bfdprofiledefault
```

Comment:
the name of the profile here is inconsistent with the name in the BGPPeer example above.